lib/puppet_x/puppetlabs/powershell/powershell_manager.rb

```diff
@@ -394,10 +394,18 @@ def read_streams
         pipe_reader = Thread.new(@pipe) do |pipe|
           # read a Little Endian 32-bit integer for length of response
           expected_response_length = pipe.sysread(4).unpack('V').first
-          return nil if expected_response_length == 0
 
-          # reads the expected bytes as a binary string or fails
-          pipe.sysread(expected_response_length)
+          if expected_response_length == 0
+            nil
+          else
+            # reads the expected bytes as a binary string or fails
+            buffer = ""
+            # Reads in the pipe data 8K, or less, at a time
+            while (buffer.length < expected_response_length)
+              buffer << pipe.sysread([expected_response_length - buffer.length, 8192].min)
```

Comment:
Interesting....

Reply:
Yeah "interesting" was not the word I was using ...

Reply:
So the problem with this is that it doesn't really solve whatever caused this problem - and it really just changes the way in which the problem may appear in the future.

What's to say that asking for 8k rather than asking for the whole thing doesn't lock the same way?

There might be a few clues in https://www.ruby-forum.com/t/sysread-changes-behavior-in-the-presence-of-threads/61176/4

It might make sense to look at the Ruby C code rather than reading the docs. If `sysread` doesn't behave the same way across platforms, it may be better to use something that's consistent.

Reply:
This is also a really inefficient way of doing allocations in Ruby and will generate a lot of garbage.

Reply:
Oh and not sure why the short circuit return was removed in favor of deeper branching?

Reply:
> .. early return was removed

You can't call `return` in that context (block).  Ruby was actually throwing errors but the Thread was swallowing them `LocalJumpError: unexpected return`

Reply:
> So the problem with this is that it doesn't really solve whatever caused this problem - and it really just changes the way in which the problem may appear in the future.

Sadly yes :-(

Reply:
I always forget that rule about the block and should have looked up the rule again before making my comment (apologies for being in a bit of a rush when I left those comments). I think we can use `break` (which returns `nil` by default) or to be more explicit `break nil`.

https://makandracards.com/makandra/46939-ruby-a-small-summary-of-what-return-break-and-next-means-for-blocks

Reply:
Nope.... `break` throws
`#<LocalJumpError: break from proc-closure>`